File: backend/app/schemas/billing.py
from datetime import date, datetime

from pydantic import BaseModel, Field

from app.models.deposit import DepositRecordStatus, DepositType
from app.models.invoice import InvoiceStatus, ReferenceType
from app.models.payment import PaymentMethod, PaymentRecordStatus
from app.models.revenue_recognition import RecognitionStatus, RecognitionType


# ── Brief models ─────────────────────────────────────────────────────────────

class CustomerBrief(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    first_name: str
    last_name: str
    company: str | None = None


class UserBrief(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    username: str
    full_name: str | None = None


class ContractBrief(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    contract_number: str
    status: str


# ── Invoice schemas ──────────────────────────────────────────────────────────

class InvoiceItemOut(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    invoice_id: int
    billing_cycle_id: int | None = None
    line_number: int
    item_code: str | None = None
    description: str
    unit: str | None = None
    quantity: float
    unit_rate: float
    amount: float
    tax_amount: float
    line_total: float
    sort_order: int
    created_at: datetime
    updated_at: datetime | None = None


class InvoiceItemCreate(BaseModel):
    item_code: str | None = Field(default=None, max_length=100)
    description: str = Field(..., min_length=1, max_length=500)
    unit: str | None = Field(default=None, max_length=50)
    quantity: float = Field(default=1.0, gt=0)
    unit_rate: float = Field(..., ge=0)


class InvoiceCreate(BaseModel):
    contract_id: int | None = None
    customer_id: int
    reference_type: ReferenceType | None = None
    reference_id: int | None = None
    issue_date: date | None = None
    due_date: date | None = None
    billing_period_start: date | None = None
    billing_period_end: date | None = None
    tax_rate: float = Field(default=0.0, ge=0, le=100)
    discount_amount: float = Field(default=0.0, ge=0)
    currency: str = Field(default="LAK", max_length=3)
    exchange_rate: float = Field(default=1.0, gt=0)
    bank_details: str | None = None
    vehicle_make: str | None = Field(default=None, max_length=100)
    vehicle_model: str | None = Field(default=None, max_length=100)
    vehicle_vin: str | None = Field(default=None, max_length=100)
    vehicle_engine_no: str | None = Field(default=None, max_length=100)
    vehicle_reg_no: str | None = Field(default=None, max_length=50)
    job_number: str | None = Field(default=None, max_length=50)
    notes: str | None = None
    internal_notes: str | None = None
    billing_cycle_ids: list[int] = Field(default_factory=list)
    items: list[InvoiceItemCreate] = Field(default_factory=list)


class InvoiceUpdate(BaseModel):
    due_date: date | None = None
    tax_rate: float | None = Field(default=None, ge=0, le=100)
    discount_amount: float | None = Field(default=None, ge=0)
    exchange_rate: float | None = Field(default=None, gt=0)
    bank_details: str | None = None
    notes: str | None = None
    internal_notes: str | None = None


class InvoiceOut(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    invoice_number: str
    contract_id: int | None = None
    customer_id: int
    reference_type: str | None = None
    reference_id: int | None = None
    status: str
    issue_date: date | None = None
    due_date: date | None = None
    paid_date: date | None = None
    subtotal: float
    tax_rate: float
    tax_amount: float
    discount_amount: float
    total_amount: float
    amount_paid: float
    balance_due: float
    currency: str
    billing_period_start: date | None = None
    billing_period_end: date | None = None
    customer: CustomerBrief
    contract: ContractBrief | None = None
    creator: UserBrief | None = None
    created_at: datetime
    updated_at: datetime | None = None
    is_active: bool


class InvoiceDetail(InvoiceOut):
    items: list[InvoiceItemOut] = []
    allocations: list["PaymentAllocationOut"] = []
    notes: str | None = None
    internal_notes: str | None = None
    sent_at: datetime | None = None
    cancelled_at: datetime | None = None
    cancellation_reason: str | None = None
    vehicle_make: str | None = None
    vehicle_model: str | None = None
    vehicle_vin: str | None = None
    vehicle_engine_no: str | None = None
    vehicle_reg_no: str | None = None
    job_number: str | None = None
    exchange_rate: float
    bank_details: str | None = None


class InvoiceListResponse(BaseModel):
    items: list[InvoiceOut]
    total: int
    page: int
    page_size: int
    pages: int


# ── Invoice workflow actions ─────────────────────────────────────────────────

class InvoiceIssueAction(BaseModel):
    issue_date: date | None = None
    due_date: date | None = None


class InvoiceSendAction(BaseModel):
    notes: str | None = None


class InvoiceCancelAction(BaseModel):
    cancellation_reason: str = Field(..., min_length=1)


class InvoiceFromCyclesRequest(BaseModel):
    contract_id: int
    billing_cycle_ids: list[int] = Field(..., min_length=1)
    tax_rate: float = Field(default=0.0, ge=0, le=100)
    discount_amount: float = Field(default=0.0, ge=0)
    currency: str = Field(default="LAK", max_length=3)
    notes: str | None = None


# ── Payment schemas ──────────────────────────────────────────────────────────

class PaymentCreate(BaseModel):
    customer_id: int
    contract_id: int | None = None
    payment_method: PaymentMethod
    amount: float = Field(..., gt=0)
    currency: str = Field(default="LAK", max_length=3)
    payment_date: date
    received_date: date | None = None
    reference_number: str | None = Field(default=None, max_length=100)
    notes: str | None = None


class PaymentUpdate(BaseModel):
    payment_method: PaymentMethod | None = None
    amount: float | None = Field(default=None, gt=0)
    payment_date: date | None = None
    received_date: date | None = None
    reference_number: str | None = Field(default=None, max_length=100)
    notes: str | None = None


class PaymentOut(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    payment_number: str
    customer_id: int
    contract_id: int | None = None
    payment_method: str
    payment_status: str
    amount: float
    currency: str
    payment_date: date
    received_date: date | None = None
    reference_number: str | None = None
    notes: str | None = None
    confirmed_by: int | None = None
    confirmed_at: datetime | None = None
    customer: CustomerBrief
    contract: ContractBrief | None = None
    confirmer: UserBrief | None = None
    creator: UserBrief | None = None
    created_at: datetime
    updated_at: datetime | None = None


class PaymentListResponse(BaseModel):
    items: list[PaymentOut]
    total: int
    page: int
    page_size: int
    pages: int


class PaymentConfirmAction(BaseModel):
    notes: str | None = None


class PaymentRejectAction(BaseModel):
    reason: str | None = None


# ── Payment Allocation schemas ───────────────────────────────────────────────

class PaymentAllocationCreate(BaseModel):
    payment_id: int
    invoice_id: int
    allocated_amount: float = Field(..., gt=0)


class PaymentAllocationOut(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    payment_id: int
    invoice_id: int
    allocated_amount: float
    allocated_at: datetime
    allocated_by: int | None = None


# ── Deposit schemas ──────────────────────────────────────────────────────────

class DepositCreate(BaseModel):
    contract_id: int
    customer_id: int
    deposit_type: DepositType
    amount: float = Field(..., gt=0)
    currency: str = Field(default="LAK", max_length=3)
    notes: str | None = None


class DepositUpdate(BaseModel):
    notes: str | None = None


class DepositOut(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    deposit_number: str
    contract_id: int
    customer_id: int
    deposit_type: str
    deposit_status: str
    amount: float
    currency: str
    received_date: date | None = None
    refund_date: date | None = None
    refund_amount: float
    forfeit_amount: float
    applied_amount: float
    payment_id: int | None = None
    refund_payment_id: int | None = None
    notes: str | None = None
    contract: ContractBrief
    customer: CustomerBrief
    creator: UserBrief | None = None
    created_at: datetime
    updated_at: datetime | None = None


class DepositListResponse(BaseModel):
    items: list[DepositOut]
    total: int
    page: int
    page_size: int
    pages: int


class DepositReceiveAction(BaseModel):
    received_date: date
    payment_id: int | None = None


class DepositRefundAction(BaseModel):
    refund_amount: float = Field(..., gt=0)
    refund_date: date
    refund_payment_id: int | None = None
    notes: str | None = None


class DepositForfeitAction(BaseModel):
    forfeit_amount: float = Field(..., gt=0)
    reason: str | None = None


class DepositApplyAction(BaseModel):
    apply_amount: float = Field(..., gt=0)
    invoice_id: int


# ── Revenue Recognition schemas ─────────────────────────────────────────────

class RevenueRecognitionCreate(BaseModel):
    contract_id: int
    invoice_id: int | None = None
    invoice_item_id: int | None = None
    recognition_type: RecognitionType
    recognition_date: date
    amount: float
    currency: str = Field(default="LAK", max_length=3)
    period_start: date | None = None
    period_end: date | None = None
    description: str | None = Field(default=None, max_length=500)
    notes: str | None = None


class RevenueRecognitionOut(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    recognition_number: str
    invoice_id: int | None = None
    invoice_item_id: int | None = None
    contract_id: int
    recognition_type: str
    recognition_status: str
    recognition_date: date
    amount: float
    currency: str
    period_start: date | None = None
    period_end: date | None = None
    description: str | None = None
    notes: str | None = None
    creator: UserBrief | None = None
    created_at: datetime
    updated_at: datetime | None = None


class RevenueRecognitionListResponse(BaseModel):
    items: list[RevenueRecognitionOut]
    total: int
    page: int
    page_size: int
    pages: int


# ── Dashboard ────────────────────────────────────────────────────────────────

class BillingDashboardSummary(BaseModel):
    total_invoiced: float
    total_paid: float
    total_outstanding: float
    total_overdue: float
    invoice_count: int
    payment_count: int
    currency: str = "LAK"


# Forward reference rebuild
InvoiceDetail.model_rebuild()

File: frontend/src/types/billing.ts
export type InvoiceStatus = 'draft' | 'issued' | 'sent' | 'partially_paid' | 'paid' | 'overdue' | 'cancelled' | 'voided'
export type PaymentMethod = 'cash' | 'bank_transfer' | 'check' | 'credit_card' | 'mobile_payment' | 'other'
export type PaymentRecordStatus = 'pending' | 'confirmed' | 'rejected' | 'refunded'
export type DepositType = 'security' | 'advance' | 'guarantee'
export type DepositRecordStatus = 'pending' | 'received' | 'partially_refunded' | 'refunded' | 'forfeited' | 'applied'
export type RecognitionType = 'rental_income' | 'service_fee' | 'penalty_fee' | 'damage_recovery' | 'deposit_forfeiture'
export type RecognitionStatus = 'scheduled' | 'recognized' | 'reversed'
export type ReferenceType = 'work_order' | 'rental'

export interface CustomerBrief { id: number; first_name: string; last_name: string; company: string | null }
export interface UserBrief { id: number; username: string; full_name: string | null }
export interface ContractBrief { id: number; contract_number: string; status: string }

export interface InvoiceItemOut {
  id: number; invoice_id: number; billing_cycle_id: number | null
  line_number: number; item_code: string | null; description: string; unit: string | null
  quantity: number; unit_rate: number; amount: number; tax_amount: number; line_total: number
  sort_order: number; created_at: string; updated_at: string | null
}

export interface PaymentAllocationOut {
  id: number; payment_id: number; invoice_id: number
  allocated_amount: number; allocated_at: string; allocated_by: number | null
}

export interface InvoiceOut {
  id: number; invoice_number: string; contract_id: number | null; customer_id: number
  reference_type: ReferenceType | null; reference_id: number | null
  status: InvoiceStatus; issue_date: string | null; due_date: string | null; paid_date: string | null
  subtotal: number; tax_rate: number; tax_amount: number; discount_amount: number
  total_amount: number; amount_paid: number; balance_due: number; currency: string
  billing_period_start: string | null; billing_period_end: string | null
  customer: CustomerBrief; contract: ContractBrief | null; creator: UserBrief | null
  created_at: string; updated_at: string | null; is_active: boolean
}

export interface InvoiceDetail extends InvoiceOut {
  items: InvoiceItemOut[]
  allocations: PaymentAllocationOut[]
  notes: string | null; internal_notes: string | null
  sent_at: string | null; cancelled_at: string | null; cancellation_reason: string | null
  vehicle_make: string | null; vehicle_model: string | null; vehicle_vin: string | null
  vehicle_engine_no: string | null; vehicle_reg_no: string | null; job_number: string | null
  exchange_rate: number; bank_details: string | null
}

export interface InvoiceListResponse { items: InvoiceOut[]; total: number; page: number; page_size: number; pages: number }

export interface PaymentOut {
  id: number; payment_number: string; customer_id: number; contract_id: number | null
  payment_method: string; payment_status: PaymentRecordStatus
  amount: number; currency: string; payment_date: string; received_date: string | null
  reference_number: string | null; notes: string | null
  confirmed_by: number | null; confirmed_at: string | null
  customer: CustomerBrief; contract: ContractBrief | null
  confirmer: UserBrief | null; creator: UserBrief | null
  created_at: string; updated_at: string | null
}

export interface PaymentListResponse { items: PaymentOut[]; total: number; page: number; page_size: number; pages: number }

export interface DepositOut {
  id: number; deposit_number: string; contract_id: number; customer_id: number
  deposit_type: string; deposit_status: DepositRecordStatus
  amount: number; currency: string
  received_date: string | null; refund_date: string | null
  refund_amount: number; forfeit_amount: number; applied_amount: number
  payment_id: number | null; refund_payment_id: number | null
  notes: string | null
  contract: ContractBrief; customer: CustomerBrief; creator: UserBrief | null
  created_at: string; updated_at: string | null
}

export interface DepositListResponse { items: DepositOut[]; total: number; page: number; page_size: number; pages: number }

export interface RevenueRecognitionOut {
  id: number; recognition_number: string
  invoice_id: number | null; invoice_item_id: number | null; contract_id: number
  recognition_type: string; recognition_status: RecognitionStatus
  recognition_date: string; amount: number; currency: string
  period_start: string | null; period_end: string | null
  description: string | null; notes: string | null
  creator: UserBrief | null; created_at: string; updated_at: string | null
}

export interface RevenueRecognitionListResponse { items: RevenueRecognitionOut[]; total: number; page: number; page_size: number; pages: number }

export interface BillingDashboardSummary {
  total_invoiced: number; total_paid: number; total_outstanding: number
  total_overdue: number; invoice_count: number; payment_count: number; currency: string
}

export interface InvoiceItemCreate { item_code?: string; description: string; unit?: string; quantity: number; unit_rate: number }

export interface InvoiceCreate {
  contract_id?: number; customer_id: number
  reference_type?: ReferenceType; reference_id?: number
  issue_date?: string; due_date?: string
  billing_period_start?: string; billing_period_end?: string
  tax_rate?: number; discount_amount?: number; currency?: string
  exchange_rate?: number; bank_details?: string
  vehicle_make?: string; vehicle_model?: string; vehicle_vin?: string
  vehicle_engine_no?: string; vehicle_reg_no?: string; job_number?: string
  notes?: string; internal_notes?: string
  billing_cycle_ids?: number[]; items?: InvoiceItemCreate[]
}
export interface InvoiceUpdate { due_date?: string; tax_rate?: number; discount_amount?: number; exchange_rate?: number; bank_details?: string; notes?: string; internal_notes?: string }
export interface InvoiceFromCyclesRequest { contract_id: number; billing_cycle_ids: number[]; tax_rate?: number; discount_amount?: number; currency?: string; notes?: string }

// Payload handed from the rental contract editor to the invoice editor via
// router state, for the Contract -> Invoice hand-off.
export interface InvoiceConversionPrefill {
  contract_id: number
  contract_number: string
  customer_id: number
  currency: string
  tax_rate: number
  items: InvoiceItemCreate[]
}
export interface PaymentCreate { customer_id: number; contract_id?: number; payment_method: PaymentMethod; amount: number; currency?: string; payment_date: string; received_date?: string; reference_number?: string; notes?: string }
export interface DepositCreate { contract_id: number; customer_id: number; deposit_type: DepositType; amount: number; currency?: string; notes?: string }

export interface InvoiceListParams { q?: string; status?: string; customer_id?: number; contract_id?: number; is_active?: boolean; issue_from?: string; issue_to?: string; due_from?: string; due_to?: string; page?: number; page_size?: number; sort?: string; order?: 'asc' | 'desc' }

export interface PaymentListParams { q?: string; customer_id?: number; contract_id?: number; payment_status?: string; payment_method?: string; date_from?: string; date_to?: string; page?: number; page_size?: number; sort?: string; order?: 'asc' | 'desc' }
